### ISSUE TITLE
[Documentation] "`@beartype` + `pytest` ftw!!!" — Smart Guy @Jasha10

### ISSUE THREAD

**Issue by leycec** [beartype/beartype#200]
### Discussed in https://github.com/beartype/beartype/discussions/199

<div type='discussions-op-text'>

<sup>Originally posted by **Jasha10** December 18, 2022</sup>
The thesis here is that pytest tests and fixtures can (and should!) be decorated with `@beartype`.
Doing so makes your tests robust against mis-typed fixtures and mis-typed `@mark.parametrize` decorators.

## Motivation
Sometimes using the pytest framework is risky:

```python
from pytest import fixture

@fixture
def foo_int() -> int:
    return 123

def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
It's easy to see that the above test will fail since `123 != "abc"`. But can you spot the type error?
```text
$ pytest test_foo.py
...
FAILED test_foo.py::test_foo - AssertionError: assert 123 == 'abc'
1 failed in 0.05s
```
Running static checkers like `mypy` and `pyright` on the above code will *not* produce any errors.
... But the `foo_int` argument passed to the `test_foo` function actually doesn't have type `str`,
despite what the type annotation `foo_int: str` would lead you to believe.
I guess static typing can't solve everything.

## Enter beartype!

With the addition of a decorator, you can ensure that your mis-typed code will crash spectacularly:
```python
from beartype import beartype
from pytest import fixture

@fixture
def foo_int() -> int:
    return 123

@beartype  # Add decorator here!
def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
```text
$ pytest test_foo.py -q
...
E   beartype.roar.BeartypeCallHintParamViolation: @beartyped test_foo.test_foo() parameter foo_int=123 violates type hint <class 'str'>, as int 123 not instance of str.

<@beartype(test_foo.test_foo) at 0x7f18eac17130>:22: BeartypeCallHintParamViolation
====================== short test summary info =======================
FAILED test_foo.py::test_foo - beartype.roar.BeartypeCallHintParamViolation: @beartyped test_foo...
1 failed in 0.10s
```
The test doesn't run! We are immediately given a bear-typed slap on the wrist.
The above diagnostic info is enough to piece together that the signature of `test_foo` is mistyped.

What advantages are there to using `@beartype` with pytest? Why should I care whether my test fails with a `BeartypeCallHintParamViolation` or with a standard pytest `AssertionError` as in the first example above?
Using `@beartype`, you can follow the Single Responsibility Principle in your test code. Your test functions' body needs only to worry about test logic, not about verifying that the fixtures or `parametrize`'d arguments are correctly-typed.

## Recipes:

Beartype works with pytest tests:
```python
from beartype import beartype
@beartype  # ensure `foo_int` *actually* has type `str`
def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
Beartype works with pytest fixtures:
```python
from beartype import beartype
from pytest import fixture
@fixture
@beartype  # ensure the `foo_int` fixture *actually* returns an `int`
def foo_int() -> int:
    return 123
```
Beartype can make sure your `@mark.parametrize` arguments are up to snuff:
```python
from beartype import beartype
from pytest import mark
@mark.parametrize("arg", [1, 2, "three"])
@beartype  # Ensure that `arg` *actually* has type `int`
def test_small_ints(arg: int) -> None:
    assert arg < 4
```</div>

**Comment by leycec** [comment]
**Excellent expose!** This would be fantastic fodder for our monolithic `README.rst` that increasingly frightens even me. Specifically:

* Let's shamelessly copy-paste all three of the `pytest` recipes you've so graciously donated to the QA world above into a new `PYTEST` subsection of [our existing *Cheatsheet*](https://github.com/beartype/beartype#cheatsheet).
* Let's shamelessly copy-paste your *wonderfully* well-written introduction into a new FAQ entry on why (*"because you'd be a blatant fool not to!"*), when (*always. just do it always.*), and how (which is kinda non-trivial, because order of decorations is significant) to integrate @beartype with `pytest`.

Thanks so much for your superbly helpful write-up, @Jasha10! I especially :heart_eyes: reading soothing words like this:

> I guess static typing can't solve everything.

<img src="https://media.tenor.com/dB61WWrIAY4AAAAS/gif.gif" width="800">

**Comment by foxx** [comment]
I'm almost convinced that @leycec is a clairvoyant, as I was _just_ about to raise a bug report for this very thing, although my apologies in advance if this is an unrelated.

I have noticed a weird issue when trying to use `beartype` with `pytest_async.fixture`. It seems to change the return/yield type of the class method it's decorating, which I've never encountered before.

Below is an example, please excuse any errors as it's a copy/paste which has been stripped to only it's relevant parts.

```python
@pytest_asyncio.fixture
@beartype
async def fixture_with_beartype(self):
    """."""
    yield dict()

async def test_fixture(self, fixture_with_beartype):
    """."""
    assert False, (fixture_with_beartype, type(fixture_with_beartype))
```

This results in the following exception, which shows the return type is wrapped as an `async_generator dict`.

```python
AssertionError: (
    <async_generator object dict at 0x7f8be6e9e4c0>, 
    <class 'async_generator'>
)
```

However, if I remove beartype decorator, the return type changes back to `dict`, e.g.

```python
@pytest_asyncio.fixture
async def fixture_without_beartype(self):
    """."""
    yield dict()

async def test_fixture(self, fixture_without_beartype):
    """."""
    assert False, (fixture_without_beartype, type(fixture_without_beartype))
```

```python
AssertionError: ({}, <class 'dict'>)
```

**Comment by leycec** [comment]
**Ah-ha!** A treacherous spiny-haired `asyncio` bug rears up on fearsome chitinous appendages while slavering disgusting green spittle everywhere. Someone, fetch my shoe.

By which I mean to say... `@pytest_asyncio.fixture` *might* be busted here. @beartype might not actually be to blame. Maybe. `asyncio` is a wicked miasma of confusion under the best of times. We're intermingling @beartype, `pytest-asyncio`, and asynchronous generators here. This is surely the worst of times.

There are a few suspicious code smells I smell all the way from Canada:

* **@beartype reduces to a noop here.** `fixture_without_beartype` isn't actually annotated by type hints. @beartype slyly notices that and then promptly does nothing by just returning `fixture_without_beartype` unmodified. Let's manually verify this for the sake of Santa Bearclaws:

```python
>>> from beartype import beartype
>>> async def fixture_with_beartype(self):
...     """."""
...     yield dict()
>>> fixture_with_beartype is beartype(fixture_with_beartype)
True  # <-- Rejoice with sweet maple syrup imported from Canada!
```

* **The `@pytest_asyncio.fixture` decorator seems to think that @beartype is doing something here.** But... we have manually verified above that @beartype is not, indeed, doing anything here! Therefore, we conclude that `@pytest_asyncio.fixture` is dumb. <sup> :grin: </sup>

I suspect that either **(A)** I actually have no idea what I am talking about *or* **(B)** `@pytest_asyncio.fixture` is doing something unholy and profane to detect and differentiate coroutines from asynchronous generators.

Also, is `@pytest_asyncio.fixture` even doing the right thing here? Surely nobody wants `@pytest_asyncio.fixture` to strip asynchronous behaviour from coroutines and asynchronous generators... do they? Doesn't that defeat the point of going `async` in the first place? Is this *really* what everyone wants?

> However, if I remove @beartype decorator, the return type changes back to `dict`.

Really? That's what you want? If so, that's fascinating. I should confess here that @beartype [internally prefers its own hand-rolled `pytest` plugin for enabling asynchronous tests](https://github.com/beartype/beartype/blob/c449dfc82d4f55e6f396693a96442ecd4070e5a6/beartype_test/conftest.py#L76). I've never actually used `pytest-asyncio`, because I just copy-paste the above code into every *other* Python project that needs asynchronous tests.

Therefore, we conclude that I have no idea what I am talking about.

**Comment by foxx** [comment]
Thanks for the detailed breakdown, that was super helpful!

I would agree that it's looking like `pytest-asyncio` is the culprit here. Admittedly I opened up the source code for `pytest-asyncio` and almost cried, there's so much magic it's making my head spin.

I'll probably create a standalone repository that reproduces this bug, along with some stack dump comparisons with and without beartype. I've been meaning to write some tooling for this sort of debugging anyway, e.g. dump stack on various breakpoints and compare differences, but needs some automation to strip out memory addresses before diff etc. Either way, it gives me an excuse to go diving into Python internals for some fun and profit :>

For now I think we can consider this an upstream issue and I'll report back my findings either way.

**Comment by leycec** [comment]
> Admittedly I opened up the source code for `pytest-asyncio` and almost cried, there's so much magic it's making my head spin.

**Yes, yes, so much yes!** I too openly wept salty tears into my rusty tea strainer stuffed with South African rooibos tea, which then tasted bad. It's not for lil' ol' me <sup>:relaxed: </sup> to judge other hard-working open-source projects that everybody else loves, but I shall now do just that.

`pytest-asyncio` is stupid overkill. Nobody uses 99% of that codebase, most of which is unmaintainable. That's why I delicately extracted the useful core of `pytest-asyncio` into [@beartype's own internal `pytest` plugin that extends `pytest` with asynchronous test support](https://github.com/beartype/beartype/blob/c449dfc82d4f55e6f396693a96442ecd4070e5a6/beartype_test/conftest.py#L76). Can you guess how big our plugin is? That's right. You clearly guessed correctly, because you are smarter than me: **17 lines.** That's it. `pytest-asyncio` really only contains 17 lines of useful code. The rest is mostly just code debt, bit rot, and salty tears.

I pontificated at great length on the `pytest` issue tracker about how `pytest` itself should just ship out-of-the-box support for asynchronous tests, because it only takes 17 lines of code to do so.
`pytest` developers told me to go away. Thus another fragment of my youthful idealism died. :sob: 

> I'll probably create a standalone repository that reproduces this bug, along with some stack dump comparisons with and without beartype.

**OMG.** You work so hard on behalf of the QA world, @foxx. If @beartype is at fault, I'll chain myself to my Star Trek-like Kinesis keyboard until I fix this. Either way, thanks so much for digging deep into yet another brutal foray into the sweaty bowels of Python testing.

**Comment by tusharsadhwani** [comment]
Heard about beartype a few minutes ago, and if this is the kind of discussions I can expect to have as a user, sign me up! Crazy that I've used mypy for a long time (and hoping to write a book on it) and hadn't heard about beartype until now. I have to go back and refactor beartype into a bunch of my notes now.

**Comment by tusharsadhwani** [comment]
FWIW this is a great usecase, I was hoping pytest would ship a mypy plugin to test its fixtures but it doesn't, but this is an equally good solution.

**Comment by leycec** [comment]
> Heard about beartype a few minutes ago

![welcome](https://media1.tenor.com/m/T8-Tt-1w-9gAAAAC/mst3k-join-us.gif)

> ...and if this is the kind of discussions I can expect to have as a user, sign me up!

**It is!** Oh, how it is. Here at Santa Bear Claws Inc., we promise mutilated bugs for Christmas *or* all your GitHub stars back.

> ...and hadn't heard about beartype until now.

No one hears about @beartype until their [latest LLaMa](https://ai.meta.com/blog/large-language-model-llama-meta-ai/) crashes in unquenchable flames around the ankles of the obnoxious company mascot, at which point it's far too late. When the mascot pathetically collapses, even @beartype can no longer help.

![a bridge too far](https://media1.tenor.com/m/u46xy1ssQu8AAAAd/raptors-fail.gif)

> FWIW this is a great usecase...

**Indeed!** Praise be to @Jasha10. He is the Smart Guy.

This documentation request also dovetails nicely into our [now oldest feature request for an official `pytest-beartype` plugin](https://github.com/beartype/beartype/issues/45). Thanks to the [terrifying power of beartype import hooks](https://beartype.readthedocs.io/en/latest/api_claw), it should now be trivial to automate away *all* `@beartype` + `pytest` integration with a single one-liner.

Ideally, nobody should have to manually decorate tests and fixtures with `@beartype` anymore. Instead, everybody should just install `pytest-beartype`. Naturally, `pytest-beartype` doesn't exist. B-b-but... it *could*. Somebody, make `pytest-beartype`.

That's all I want for Christmas this year. That and lots of video games.

**Comment by tusharsadhwani** [comment]
It's decideed then. I'm making `pytest-beartype`.

**Comment by leycec** [comment]
**Wondrousness exemplified.** As in all things, I stand on the shoulders of giants. You are that giant. So is `typeguard` – which, conveniently, already implemented a [comparable `pytest` plugin probably *decades* ago](https://github.com/agronholm/typeguard/blob/master/src/typeguard/_pytest_plugin.py).

And... it's actually pretty trivial. Well, this is awkward. Kinda ashamed this never happened earlier, now that I'm shamelessly stripping everything that's not nailed down from `typeguard`. In theory, something resembling the following *should* suffice:

```python
def pytest_addoption(parser: 'pytest.Parser') -> None:

    # Add beartype-specific "pytest" options exposed by this plugin.
    group = parser.getgroup('beartype')
    group.addoption(
        '--beartype-packages',
        action='store',
        help=(
            'comma-delimited list of the fully-qualified names of '
            'all packages and modules to type-check with beartype'
        ),
    )


def pytest_configure(config: 'pytest.Config') -> None:

    # Comma-delimited string listing the fully-qualified names of *ALL* packages
    # and modules to type-check with beartype, corresponding to the
    # "--beartype-packages" option defined above by the pytest_addoption() hook.
    package_names_str = config.getoption('beartype_packages')

    # If the user passed this option...
    if package_names_str:
        # Defer hook-specific imports. To improve "pytest" startup performance,
        # avoid performing *ANY* imports unless the user actually passed the
        # "--beartype-packages" option declared by this plugin.
        from beartype.roar import BeartypeWarning
        from beartype.claw import beartype_package
        from beartype._util.text.utiltextjoin import join_delimited
        from sys import modules
        from warnings import warn

        class BeartypePytestWarning(BeartypeWarning):
            '''
            Beartype :mod:`pytest` warning.

            This warning is emitted at :mod:`pytest` configuration time when one or
            more packages or modules to be type-checked have already been imported
            under the active Python interpreter.
            '''

            pass

        # Tuple of the fully-qualified names of these packages and modules.
        package_names = tuple(
            package_name_str.strip()
            for package_name_str in package_names_str.split(',')
        )

        # Tuple of the subset of these names corresponding to previously
        # imported packages and modules under the active Python interpreter.
        package_imported_names = tuple(
            package_name
            for package_name in package_names
            if package_name in modules
        )

        # If at least one of these packages or modules has already been
        # imported...
        if package_imported_names:
            # Comma-delimited double-quoted string listing these packages. Yeah!
            package_imported_names_str = join_delimited(
                strs=already_imported_packages,
                delimiter_if_two=' and ',
                delimiter_if_three_or_more_nonlast=', ',
                delimiter_if_three_or_more_last=', and ',
                is_double_quoted=True,
            )

            # Emit a non-fatal warning informing the user.
            warn(
                (
                    f'Previously imported packages '
                    f'{package_imported_names_str} not checkable by beartype.'
                ),
                BeartypePytestWarning,
                stacklevel=1,  # <-- dark magic glistens dangerously
            )

        # Install an import hook type-checking these packages and modules.
        beartype_package(package_names)
```

No biggie. You know. Now, I know this sounds cray-cray. We've only just met. B-b-but... would you like to be hosted under the official @beartype organization at `https://github/beartype/pytest-beartype`? If so, please feel free to **temporarily** <sup>i promise, would an internet stranger lie!?</sup> transfer ownership over the `pytest-beartype` GitHub repository to the @beartype organization. I will then **immediately** and **permanently** <sup>i promise x 2 – if u can't trust internet strangers, who can u trust??</sup> add you back as an administrator. I solemnly swear I am a nice guy who lives in a cabin in the woods.

GitHub mandates this awkward workflow, sadly. Used to be easier and safer for everybody back in the day. *Anyways!* Hype and excitement grows. Team Bear Bros: unite for the common good of QA. :muscle: :bear:

**Comment by tusharsadhwani** [comment]
GitHub does not like that idea.

![image](https://github.com/beartype/beartype/assets/43412083/48719f58-c2dc-4e46-aab0-7106b2a1c5e8)

**Comment by tusharsadhwani** [comment]
Good morning, non sleepy Tushar speaking now, I think I need to be a member of the org before I can transfer? I'm not very sure still.

**Comment by leycec** [comment]
This is sleepy @leycec speaking on a frosty Friday night from the wintery bowels of Canada. Several decades ago, GitHub used to be easier. Now, I am old and bald and the modern GitHub UX workflow makes me pound my skinny fists on the grotty carpet. By which I mean to say...

**You're in!** I've fired off the invitation. And thanks so much for being so enthusiastically game for all of this. I'm already salivating at the glorious prospects of a nascent `pytest-beartype` plugin. Finally! @beartype and `pytest`, best QA bros at last. :yum:

**Comment by tusharsadhwani** [comment]
I will also happily be the first enterprise user and stress tester of `pytest-beartype`, by forcing it on my peers at my workplace. Looking forward to it.

**Comment by tusharsadhwani** [comment]
Repository transferred. Expect the initial release by the end of the weekend hopefully.

**Comment by leycec** [comment]
> ...by forcing it on my peers at my workplace.

![](https://media1.tenor.com/m/R_Q4qZqABSwAAAAC/dance-dancing.gif)
<sup>dancing grizzly bear approves of beneficial workplace extortion</sup>

> Repository transferred.

The Seventh Wonder of the QA world emerges. For my part, I've:

* **Promoted you back to Admin.** You are now the ultimate power.
* [**Pinned `pytest-beartype` in the @beartype organization.**](https://github.com/beartype) The world has now become aware of our timely enterprise.

Things that have yet to be done include:

* Publicly advertise `pytest-beartype` in [@beartype's official FAQ](https://beartype.readthedocs.io/en/latest/faq). I'm contemplating a new FAQ entry promoting `pytest-beartype` with a title like: "How do I type-check only while running tests?"

I am probably forgetting other things, because it is Saturday night and those video games aren't gonna play themselves.

> Expect the initial release by the end of the weekend hopefully.

Your productivity shames me. Please, though; we're increasingly into the holidays. There's *absolutely* no rush. This is something everyone has gone without for literally years. Everyone can go without it a bit longer.

Or just push your family to the wayside and tell them that something more important has come up. "Sorry, folks. Gotta hack on core Python infrastructure over here. Santa's gonna be late this year."

**Comment by tusharsadhwani** [comment]
Quick note, this hasn't been forgotten, I just got reminded that I had a lot of work pending from before the holidays.
I'm hoping to push some code (and hopefully tests) tonight. No takesies backsies.

**Comment by tusharsadhwani** [comment]
So you were right, that initial piece of code (apart from a tiny typo) seems to be just working.
So I've taken the liberty of [releasing `pytest-beartype` v0.0.1](https://pypi.org/p/pytest-beartype) with essentially no proper testing. Feel free to check how the developer experience is with it right now, or break it (which it will).

**Comment by leycec** [comment]
> So you were right, that initial piece of code (apart from a tiny typo) seems to be just working.

**Wondrousness!** The bear roars with pleasure. Thanks so much for putting everything through its paces *and* resolving the shameful typo that I now blame on somebody else. You're stunning.

![](https://media1.tenor.com/m/GAwX3Fzjl20AAAAC/happy-dance-bear.gif)
<sup>this bear in our driveway, it dances for you</sup>

> So I've taken the liberty of [releasing pytest-beartype v0.0.1](https://pypi.org/p/pytest-beartype) with essentially no proper testing.

**Wondrousness x 2!** Sometimes, this is the way of QA: a shot in the dark for the good of humanity. What could possibly go wrong? Yet again, thanks so much for taking on this gut-pounding responsibility despite your excruciating workload. You amaze.

> break it (which it will)

:smile: 

> In the meantime I'll be reading up on how people test pytest plugins.

I... always kinda wondered that myself. My suspicion is that they just don't; like you, everybody just shotguns it. That would explain `pytest-asyncio`, the ubiquitous `pytest` plugin that broke the test suites of both @beartype and Dask so hard that we now unconditionally disable `pytest-asyncio` in our `tox` and GitHub Actions workflows. *But we don't talk about that.*

QA-on-QA: when your QA process goes recursive, just give up.

## @beartype + `pytest-beartype` = The Power of Utopia

**Relatedly,** I'm releasing the epochal @beartype 0.17.0 in a few days. The official announcement on our forums will giddily namedrop you and `pytest-beartype`, @tusharsadhwani. Take all the GitHub stars!

If you'd ever like @beartype to public plug|rep|push|promote your up-coming New York Hacker News bestseller *or* anything else you've got spinning on your plate, please let me know. @beartype is here to feed you fame and money. Okay, mostly fame. Ain't nobody got money bags in the Canadian wilderness. :money_mouth_face: :money_with_wings: :moneybag: <sup>*gods why are there so many money emoji*</sup>

Lastly, let's leave this issue quietly open until somebody else adds a FAQ entry promoting `pytest-beartype`. At long last, the puzzle is almost complete.

**Comment by leycec** [comment]
**Resolved by ae20e16.** Praise be to @tusharsadhwani. He saved @beartype from @Leycec's ineptitude. First, he made `pytest-beartype`. Then, he made @beartype proudly advertise `pytest-beartype` in a new FAQ entry. @tusharsadhwani is truly an unstoppable juggernaut of QA.

All is now right with the world. Thus, this issue now closes itself.
